### PR TITLE
KAFKA-2706: make state stores first class citizens in the processor topology

### PR DESCRIPTION
**Pull request by ymatsuda**
- Added StateStoreSupplier
- StateStore
  - Added init(ProcessorContext context) method
- TopologyBuilder
  - Added addStateStore(StateStoreSupplier supplier, String... processNames)
  - Added connectProessorAndStateStores(String processorName, String... stateStoreNames)
    - This is for the case processors are not created when a store is added to the topology. (used by KStream)
- KStream
  - add stateStoreNames to process(), transform(), transformValues().
- Refactored existing state stores to implement StateStoreSupplier

@guozhangwang

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java [hunk 1]
```diff
@@ -241,20 +248,81 @@ public final TopologyBuilder addProcessor(String name, ProcessorSupplier supplie
...
+        if (!nodeFactories.containsKey(processorName))
+            throw new TopologyException("Processor " + processorName + " is not added yet.");
+
+        Set<String> users = stateStoreUsers.get(stateStoreName);
+        Iterator<String> iter = users.iterator();
+        if (iter.hasNext()) {
+            String user = iter.next();
+            nodeGrouper.unite(user, processorName);
+        } else {
+            users.add(processorName);
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java]: It seems we will only have one recorded user inside each store's user set?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTestDriver.java [hunk 2]
```diff
@@ -14,7 +14,7 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.kafka.test;
+package org.apache.kafka.streams.processor.internals;
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTestDriver.java]: Why do we want to move it into processor.internals package?

File: streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java (hunk 1 above)
Comment (ymatsuda) [inline streams/src/main/java/org/apache/kafka/streams/processor/TopologyBuilder.java]: oops... I will fix it.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTestDriver.java (hunk 2 above)
Comment (ymatsuda) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/ProcessorTopologyTestDriver.java]: I added a package local method ProcessorContextImpl.getStateMgr() for a test to use. The state manager is our internal object so I don't want the method to be public. For a test code to use this method it has to be in the same package.

**Comment by ymatsuda** [comment]
Jenkins failures seem to be unrelated to this change. @guozhangwang

**Comment by guozhangwang** [comment]
Thanks, LGTM.

**Pull request closed without merging**